`depsolvednf` unit tests don't PASS with DNF5 solver

Downstream Jira issue: https://issues.redhat.com/browse/HMS-10324

The osbuild repository runs the `depsolvednf` unit tests against its version of the depsolver. There is a CI test that can run the unit tests with DNF4 and DNF5, [but the DNF5 version is currently disabled](https://github.com/osbuild/osbuild/blob/21e6baf7a83873bcb45202bedddaebf534d144d4/.github/workflows/test.yml#L121-L122), because the tests are failing. Specifically:

One can simply run the following command in an osbuild repository checkout:
```shell
./test/cases/images-depsolve-dnf-tests --images-ref "main" --use-dnf5
```

You'll get the following errors:
```
=== RUN   TestErrorRepoInfo
=== RUN   TestErrorRepoInfo/V2
=== RUN   TestErrorRepoInfo/V2/0

(process:82152): librepo-WARNING **: 13:50:27.457: LRO_METALINKURL processing failed: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
RepoError: There was a problem reading a repository: Failed to download metadata (metalink: "https://0.0.0.0/baseos/metalink") for repository "446aa191a2ffd36e7fb2caaeb056b3921083671e8b9e4aae898d2d9002db5939": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
=== NAME  TestErrorRepoInfo/V2
    depsolvednf_test.go:986: 
        	Error Trace:	/tmp/tmp0hmi6ql_/images/pkg/depsolvednf/depsolvednf_test.go:986
        	Error:      	"DNF error occurred: RepoError: There was a problem reading a repository: Failed to download metadata (metalink: \"https://0.0.0.0/baseos/metalink\") for repository \"446aa191a2ffd36e7fb2caaeb056b3921083671e8b9e4aae898d2d9002db5939\": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]" does not contain "https://0.0.0.0/baseos/repo"
        	Test:       	TestErrorRepoInfo/V2
=== RUN   TestErrorRepoInfo/V2/1

(process:82154): librepo-WARNING **: 13:50:27.556: LRO_METALINKURL processing failed: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
RepoError: There was a problem reading a repository: Failed to download metadata (metalink: "https://0.0.0.0/baseos/metalink") for repository "446aa191a2ffd36e7fb2caaeb056b3921083671e8b9e4aae898d2d9002db5939": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
=== NAME  TestErrorRepoInfo/V2
    depsolvednf_test.go:986: 
        	Error Trace:	/tmp/tmp0hmi6ql_/images/pkg/depsolvednf/depsolvednf_test.go:986
        	Error:      	"DNF error occurred: RepoError: There was a problem reading a repository: Failed to download metadata (metalink: \"https://0.0.0.0/baseos/metalink\") for repository \"446aa191a2ffd36e7fb2caaeb056b3921083671e8b9e4aae898d2d9002db5939\": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]" does not contain "https://0.0.0.0/baseos/repo"
        	Test:       	TestErrorRepoInfo/V2
=== RUN   TestErrorRepoInfo/V2/2

(process:82156): librepo-WARNING **: 13:50:27.661: LRO_METALINKURL processing failed: Curl error (7): Could not connect to server for https://0.0.0.0/f35/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
RepoError: There was a problem reading a repository: Failed to download metadata (metalink: "https://0.0.0.0/f35/metalink") for repository "2d021dc3a4cd71ba24167000ebb593c063cc321e6eca1139d2060557e882ee8f": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/f35/metalink [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
=== RUN   TestErrorRepoInfo/V2/3

(process:82158): librepo-WARNING **: 13:50:27.765: LRO_MIRRORLISTURL processing failed: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/mirrors [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
RepoError: There was a problem reading a repository: Failed to download metadata (mirrorlist: "https://0.0.0.0/baseos/mirrors") for repository "c38202787e185186f8eaf72839024f7b07a4c6e54c97061cbee16fd313a68e8f": Cannot prepare internal mirrorlist: Curl error (7): Could not connect to server for https://0.0.0.0/baseos/mirrors [Failed to connect to 0.0.0.0 port 443 after 0 ms: Could not connect to server]
--- FAIL: TestErrorRepoInfo (0.41s)
    --- FAIL: TestErrorRepoInfo/V2 (0.41s)
        --- PASS: TestErrorRepoInfo/V2/0 (0.10s)
        --- PASS: TestErrorRepoInfo/V2/1 (0.10s)
        --- PASS: TestErrorRepoInfo/V2/2 (0.11s)
        --- PASS: TestErrorRepoInfo/V2/3 (0.10s)
```

and

```
=== RUN   TestSolverDepsolveAll/V2/multi-chain-error-first
libdnf5::OptionChild<libdnf5::OptionBool>& libdnf5::repo::ConfigRepo::get_gpgcheck_option() is deprecated: Use get_pkg_gpgcheck_option()
error install_specs
MarkingErrors: Error occurred when marking packages for installation: No match for argument: does-not-exist
    depsolvednf_test.go:448: 
        	Error Trace:	/tmp/tmp0hmi6ql_/images/pkg/depsolvednf/depsolvednf_test.go:448
        	Error:      	"error depsolving package sets for \"first\": DNF error occurred: MarkingErrors: Error occurred when marking packages for installation: No match for argument: does-not-exist" does not contain "missing packages: does-not-exist"
        	Test:       	TestSolverDepsolveAll/V2/multi-chain-error-first
=== RUN   TestSolverDepsolveAll/V2/multi-chain-error-second
libdnf5::OptionChild<libdnf5::OptionBool>& libdnf5::repo::ConfigRepo::get_gpgcheck_option() is deprecated: Use get_pkg_gpgcheck_option()
libdnf5::OptionChild<libdnf5::OptionBool>& libdnf5::repo::ConfigRepo::get_gpgcheck_option() is deprecated: Use get_pkg_gpgcheck_option()
error install_specs
MarkingErrors: Error occurred when marking packages for installation: No match for argument: does-not-exist
    depsolvednf_test.go:448: 
        	Error Trace:	/tmp/tmp0hmi6ql_/images/pkg/depsolvednf/depsolvednf_test.go:448
        	Error:      	"error depsolving package sets for \"second\": DNF error occurred: MarkingErrors: Error occurred when marking packages for installation: No match for argument: does-not-exist" does not contain "missing packages: does-not-exist"
        	Test:       	TestSolverDepsolveAll/V2/multi-chain-error-second
=== RUN   TestSolverDepsolveAll/V2/multi-chain-with-sbom
libdnf5::OptionChild<libdnf5::OptionBool>& libdnf5::repo::ConfigRepo::get_gpgcheck_option() is deprecated: Use get_pkg_gpgcheck_option()
libdnf5::OptionChild<libdnf5::OptionBool>& libdnf5::repo::ConfigRepo::get_gpgcheck_option() is deprecated: Use get_pkg_gpgcheck_option()
--- FAIL: TestSolverDepsolveAll (3.47s)
    --- FAIL: TestSolverDepsolveAll/V2 (3.47s)
        --- PASS: TestSolverDepsolveAll/V2/flat (0.49s)
        --- PASS: TestSolverDepsolveAll/V2/chain (0.41s)
        --- PASS: TestSolverDepsolveAll/V2/multi-chain (0.80s)
        --- FAIL: TestSolverDepsolveAll/V2/multi-chain-error-first (0.19s)
        --- FAIL: TestSolverDepsolveAll/V2/multi-chain-error-second (0.58s)
        --- PASS: TestSolverDepsolveAll/V2/multi-chain-with-sbom (1.00s)
```

It seems that DNF5 behavior slightly differs from DNF4. These unit tests need to be investigated and ideally modified to work and pass with both, DNF4 and DNF5 version of the solver.

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -65,6 +65,12 @@ jobs:
       - name: Run depsolver tests with force-dnf to make sure it's not skipped
         run: go test -race ./pkg/depsolvednf/... -force-dnf
 
+      - name: Run depsolver tests with dnf5
+        run: |
+          sudo sed -i 's/"use_dnf5": false/"use_dnf5": true/' /usr/lib/osbuild/solver.json
+          go clean -testcache
+          go test -race ./pkg/depsolvednf/... -force-dnf
+
   container-resolver-tests:
     name: "🛃 Container resolver tests"
     runs-on: ubuntu-24.04

--- a/.github/workflows/tests.yml
+++ b/.github/workflows/tests.yml
@@ -60,16 +60,15 @@ jobs:
       - name: Run unit tests
         # Skip tests that resolve manifests from local container storage.
         # They are tested separately (see below: requires root)
-        run: go test  -race -timeout 20m ./... -test.skip 'TestBlockingResolverLocalManifest|TestResolverLocalManifest|TestGetManifestLocal'
+        run: go test -count=1 -race -timeout 20m ./... -test.skip 'TestBlockingResolverLocalManifest|TestResolverLocalManifest|TestGetManifestLocal'
 
       - name: Run depsolver tests with force-dnf to make sure it's not skipped
-        run: go test -race ./pkg/depsolvednf/... -force-dnf
+        run: go test -count=1 -race ./pkg/depsolvednf/... -force-dnf
 
       - name: Run depsolver tests with dnf5
         run: |
           sudo sed -i 's/"use_dnf5": false/"use_dnf5": true/' /usr/lib/osbuild/solver.json
-          go clean -testcache
-          go test -race ./pkg/depsolvednf/... -force-dnf
+          go test -count=1 -race ./pkg/depsolvednf/... -force-dnf
 
   container-resolver-tests:
     name: "🛃 Container resolver tests"
@@ -103,13 +102,13 @@ jobs:
       # We need to run the test as root, since we use the root
       # containers-storage for the local resolvers
       - name: Run unit tests for container resolve
-        run: sudo go test ./pkg/container/... --force-local-resolver
+        run: sudo go test -count=1 ./pkg/container/... --force-local-resolver
 
       - name: Run unit tests for bib container
-        run: sudo go test ./pkg/bootc/... --fail-if-podman-missing
+        run: sudo go test -count=1 ./pkg/bootc/... --fail-if-podman-missing
 
       - name: Run tests for bootc container
-        run: sudo go test ./pkg/distro/bootc/... ./pkg/distro/generic/...
+        run: sudo go test -count=1 ./pkg/distro/bootc/... ./pkg/distro/generic/...
 
   unit-tests-cs:
     strategy:

--- a/cmd/image-builder/main_test.go
+++ b/cmd/image-builder/main_test.go
@@ -865,7 +865,15 @@ func TestManifestOverrideRepo(t *testing.T) {
 	testutil.CaptureStdio(t, func() {
 		err = main.Run()
 	})
-	assert.ErrorContains(t, err, "forced repo#0 xxx.abcdefgh-no-such-host.com/repo: http://xxx.abcdefgh-no-such-host.com/repo]: Cannot download repomd.xml")
+
+	// When using dnf5, the error message differs substantially. More complete
+	// error messages are already covered by the depsolvednf tests for both dnf
+	// and dnf5. Here we just make sure the repo URL appears in the error
+	// message with one of two fragments for each backend.
+	assert.ErrorContains(t, err, "xxx.abcdefgh-no-such-host.com/repo")
+	if !strings.Contains(err.Error(), "Cannot download repomd.xml") && !strings.Contains(err.Error(), "Usable URL not found") {
+		assert.Failf(t, "Unexpected error message", "Error %q does not contain either of the expected messages for dnf or dnf5", err)
+	}
 	// XXX: we should probably look into "images" here, there is a bunch
 	// of redundancy in the full error message:
 	//

--- a/pkg/depsolvednf/depsolvednf_test.go
+++ b/pkg/depsolvednf/depsolvednf_test.go
@@ -971,10 +971,10 @@ func TestErrorRepoInfo(t *testing.T) {
 			restore := mockActiveHandler(h.handler)
 			defer restore()
 
-			assert := assert.New(t)
 			solver := NewSolver("platform:f38", "38", "x86_64", "fedora-38", "/tmp/cache")
 			for idx, tc := range testCases {
 				t.Run(fmt.Sprintf("%d", idx), func(t *testing.T) {
+					assert := assert.New(t)
 					_, err := solver.Depsolve([]rpmmd.PackageSet{
 						{
 							Include:      []string{"osbuild"},

--- a/pkg/depsolvednf/depsolvednf_test.go
+++ b/pkg/depsolvednf/depsolvednf_test.go
@@ -275,6 +275,8 @@ func TestSolverDepsolveAll(t *testing.T) {
 		sbomType    sbom.StandardType
 		// slice of message fragments which are all expected to be in the error
 		expErrs []string
+		// slice of message fragments which are all expected to be in the error (when using dnf5)
+		expErrs5 []string
 	}
 
 	tmpdir := t.TempDir()
@@ -365,7 +367,8 @@ func TestSolverDepsolveAll(t *testing.T) {
 					},
 				},
 			},
-			expErrs: []string{"error depsolving package sets for \"first\"", "missing packages: does-not-exist"},
+			expErrs:  []string{"error depsolving package sets for \"first\"", "missing packages: does-not-exist"},
+			expErrs5: []string{"error depsolving package sets for \"first\"", "No match for argument: does-not-exist"},
 		},
 		"multi-chain-error-second": {
 			// two chain depsolves for different pipelines, second one fails
@@ -395,7 +398,8 @@ func TestSolverDepsolveAll(t *testing.T) {
 					},
 				},
 			},
-			expErrs: []string{"error depsolving package sets for \"second\"", "missing packages: does-not-exist"},
+			expErrs:  []string{"error depsolving package sets for \"second\"", "missing packages: does-not-exist"},
+			expErrs5: []string{"error depsolving package sets for \"second\"", "No match for argument: does-not-exist"},
 		},
 		"multi-chain-with-sbom": {
 			// two chain depsolves for different pipelines with sbom
@@ -429,6 +433,8 @@ func TestSolverDepsolveAll(t *testing.T) {
 		},
 	}
 
+	dnf5 := usingDNF5(t)
+
 	for _, h := range getTestHandlers() {
 		t.Run(h.name, func(t *testing.T) {
 			restore := mockActiveHandler(h.handler)
@@ -444,7 +450,11 @@ func TestSolverDepsolveAll(t *testing.T) {
 					res, err := solver.DepsolveAll(tc.packageSets)
 					if len(tc.expErrs) != 0 {
 						assert.Error(err)
-						for _, expErr := range tc.expErrs {
+						expErrs := tc.expErrs
+						if dnf5 {
+							expErrs = tc.expErrs5
+						}
+						for _, expErr := range expErrs {
 							assert.Contains(err.Error(), expErr)
 						}
 						return
@@ -925,12 +935,30 @@ func expectedDepsolvedPackages(repo rpmmd.RepoConfig) rpmmd.PackageList {
 	return expectedTemplate
 }
 
+func usingDNF5(t *testing.T) bool {
+	t.Helper()
+	data := map[string]bool{}
+	config, err := os.ReadFile("/usr/lib/osbuild/solver.json")
+	if err != nil {
+		t.Fatalf("failed to read solver.json: %s", err)
+	}
+	if err := json.Unmarshal(config, &data); err != nil {
+		t.Fatalf("failed to parse solver.json: %s", err)
+	}
+
+	if _, ok := data["use_dnf5"]; !ok {
+		t.Fatal("expected property 'use_dnf5' not found in solver.json")
+	}
+	return data["use_dnf5"]
+}
+
 func TestErrorRepoInfo(t *testing.T) {
 	requireDNF(t)
 
 	type testCase struct {
-		repo   rpmmd.RepoConfig
-		expMsg string
+		repo    rpmmd.RepoConfig
+		expMsg  string
+		expMsg5 string // expected message when using dnf5
 	}
 
 	testCases := []testCase{
@@ -940,7 +968,8 @@ func TestErrorRepoInfo(t *testing.T) {
 				BaseURLs: []string{"https://0.0.0.0/baseos/repo"},
 				Metalink: "https://0.0.0.0/baseos/metalink",
 			},
-			expMsg: "https://0.0.0.0/baseos/repo",
+			expMsg:  "https://0.0.0.0/baseos/repo",
+			expMsg5: "https://0.0.0.0/baseos/metalink",
 		},
 		{
 			repo: rpmmd.RepoConfig{
@@ -948,23 +977,28 @@ func TestErrorRepoInfo(t *testing.T) {
 				BaseURLs: []string{"https://0.0.0.0/baseos/repo"},
 				Metalink: "https://0.0.0.0/baseos/metalink",
 			},
-			expMsg: "https://0.0.0.0/baseos/repo",
+			expMsg:  "https://0.0.0.0/baseos/repo",
+			expMsg5: "https://0.0.0.0/baseos/metalink",
 		},
 		{
 			repo: rpmmd.RepoConfig{
 				Name:     "fedora",
 				Metalink: "https://0.0.0.0/f35/metalink",
 			},
-			expMsg: "https://0.0.0.0/f35/metalink",
+			expMsg:  "https://0.0.0.0/f35/metalink",
+			expMsg5: "https://0.0.0.0/f35/metalink",
 		},
 		{
 			repo: rpmmd.RepoConfig{
 				Name:       "",
 				MirrorList: "https://0.0.0.0/baseos/mirrors",
 			},
-			expMsg: "https://0.0.0.0/baseos/mirrors",
+			expMsg:  "https://0.0.0.0/baseos/mirrors",
+			expMsg5: "https://0.0.0.0/baseos/mirrors",
 		},
 	}
+
+	dnf5 := usingDNF5(t)
 
 	for _, h := range getTestHandlers() {
 		t.Run(h.name, func(t *testing.T) {
@@ -983,7 +1017,11 @@ func TestErrorRepoInfo(t *testing.T) {
 						},
 					}, sbom.StandardTypeNone)
 					assert.Error(err)
-					assert.Contains(err.Error(), tc.expMsg)
+					expMsg := tc.expMsg
+					if dnf5 {
+						expMsg = tc.expMsg5
+					}
+					assert.Contains(err.Error(), expMsg)
 				})
 			}
 		})

